File: erpnext/accounts/report/customer_wise_total_sales_and_expenses/customer_wise_total_sales_and_expenses.py
# Copyright (c) 2013, Bloom Stack, Inc and contributors
# For license information, please see license.txt

from __future__ import unicode_literals

from frappe import _
import frappe


def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data


def get_columns(filters):
	return [
		{
			"fieldname": "customer_name",
			"label": _("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			"width": 200
		},
		{
			"fieldname": "total_sales",
			"label": _("Total Sales"),
			"fieldtype": "Link",
			"options": "Sales Invoice",
			"width": 200
		},
		{
			"fieldname": "total_expense",
			"label": _("Total Expense"),
			"fieldtype": "Link",
			"options": "Purchase Invoice",
			"width": 200
		},
		{
			"fieldname": "net_gl",  # Net Gain/Loss
			"label": _("Net Gain/Loss"),
			"width": 200
		}
	]


def get_data(filters=None):
	data = frappe.db.sql("""
		SELECT
			c.customer_name AS customer_name,
			SUM(IFNULL(si.grand_total, 0)) AS total_sales,
			SUM(IFNULL(pi.grand_total, 0)) AS total_expense,
			SUM(IFNULL(si.grand_total, 0)) - SUM(IFNULL(pi.grand_total, 0)) AS net_gl
		FROM
			tabCustomer c
				JOIN tabSupplier s ON c.customer_name = s.supplier_name
				LEFT JOIN `tabSales Invoice` si ON si.customer = c.customer_name
				LEFT JOIN `tabPurchase Invoice` pi ON pi.supplier = c.customer_name
		GROUP BY
			c.customer_name
		""", as_dict=True)
	return data

File: erpnext/accounts/report/customer_wise_total_sales_and_expenses/customer_wise_total_sales_and_expenses.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Customer-wise Total Sales and Expenses"] = {
	"formatter": function (value, row, column, data, default_formatter) {
		if (column.fieldname == "total_sales" && data && !!data["total_sales"]) {
			value = data["total_sales"];
			column.link_onclick = "frappe.query_reports['Customer-wise Total Sales and Expenses'].set_total_sales_route_to_sales_register(" + JSON.stringify(data) + ")";
		}
		if (column.fieldname == "total_purchase" && data && !!data["total_purchase"]) {
			value = data["total_purchase"];
			column.link_onclick = "frappe.query_reports['Customer-wise Total Sales and Expenses'].set_total_expenses_route_to_purchase_register(" + JSON.stringify(data) + ")";
		}
		value = default_formatter(value, row, column, data);
		return value;
	},
	"set_total_sales_route_to_sales_register": function (data) {
		frappe.route_options = {
			"customer": data["customer_name"]
		};
		frappe.set_route("query-report", "Sales Register");
	},
	"set_total_expenses_route_to_purchase_register": function (data) {
		frappe.route_options = {
			"supplier": data["customer_name"]
		};
		frappe.set_route("query-report", "Purchase Register");
	}
};